### DISCUSSION TITLE
LSP (Rust): red underline - explanation?

### DISCUSSION THREAD

**Discussion by tf42src** [LazyVim/LazyVim#3271] [Q&A] [answered]
Trying something new .. have installed Rust ([lazyvim.plugins.extras.lang.rust](https://www.lazyvim.org/extras/lang/rust)) and the rust-analyser. Some of the code is underlined in red.

I couldn't figure out why.

![Screenshot from 2024-05-22 09-04-28](https://github.com/LazyVim/LazyVim/assets/9712861/bd4468ef-4f5b-489f-b6f9-814e25d893c5)

**Comment by folke** [comment]
probably spelling?
Did you set `vim.opt.spell = true`?

In that buffer, do `:set nospell` to see if that's the issue

**Reply by dpetka2001** [reply]
I don't think it's the spelling. It's probably because the `rust-analyzer` by Mason isn't picked up by rustaceanvim. I observed something similar in other issues. The solution was to uninstall `rust-analyzer` by Mason and install it via Rust toolchain. I have an open PR for removing the `rust-analyzer` from `nvim-lspconfig` spec for the Rust Extra, so it doesn't automatically install and the user will rely on the toolchain instead.

**Reply by tf42src** [reply]
I had installed it via `rustup component add rust-analyzer`. Does it matter how it is installed? Is there a config setting which I could point to where the executable is?

Actually it might have been installed via mason, I found it in `~/.local/share/nvim/mason/bin/rust-analyzer` (as well as in `/snap/bin/rust-analyzer`). 

I am using lazyvim which has a package manager, why do I need Mason (which is also a package manager)?

I tried uninstalling as you said using `:MasonUninstall rust-analyzer`, but lazyvim just reinstalls it.

**Reply by dpetka2001** [reply]
@tf42src Create a file in `~/.config/nvim/lua/plugins/lsp.lua` with following contents
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        rust_analyzer = {
          mason = false,
        },
      },
    },
  }
}
```
After that uninstall rust-analyzer from Mason and it shouldn't be installed again when you restart Neovim.

**Reply by tf42src** [reply]
That seems to have "worked". It doesn't reinstall, but now I don't have any formatter, analyser or anything by the looks of it.

Opening a `.rs` file yields this warning:
![Screenshot from 2024-05-23 22-36-12](https://github.com/LazyVim/LazyVim/assets/9712861/996a149d-a3fd-4617-a404-c331bbee62bb)

And running `:RustFmt` shows this error:
![Screenshot from 2024-05-23 22-37-23](https://github.com/LazyVim/LazyVim/assets/9712861/eefa7b51-ba36-4b40-b705-9cd9384447ac)

Format on save also does not work anymore. :(

**Reply by dpetka2001** [reply]
What does `:!which rust-analyzer` show from inside Neovim?

**Reply by tf42src** [reply]
`/snap/bin/rust-analyzer` - same as outside nvim

![Screenshot from 2024-05-24 23-14-57](https://github.com/LazyVim/LazyVim/assets/9712861/8d03d7fa-0a06-4b1a-91ff-bd3d48384b6f)

**Reply by dpetka2001** [reply, marked as answer]
I didn't install Rust via snap and don't know how `rustaceanvim` behaves with it. I just installed Rust as described in the [official docs](https://www.rust-lang.org/tools/install) and `rustaceanvim` works out of the box with it.

**Reply by tf42src** [reply]
Shouldn't it just use the binary it can find? It shouldn't matter how it is installed.

Either way, I installed via the official docs and now autocomplete works, but `:RustFmt` still throws an error:
![Screenshot from 2024-05-26 10-54-06](https://github.com/LazyVim/LazyVim/assets/9712861/b2291588-c494-4530-b52c-a7df5bea9ea1)

Can I get a better error message from somewhere?

**Reply by dpetka2001** [reply]
https://github.com/mrcjkb/rustaceanvim/discussions/172

**Reply by tf42src** [reply]
Thanks again!

**Comment by LingLambda** [comment]
I had the same this. The reason was that I was using the 'rust' package provided by Arch Linux. After uninstalling it and replacing it with 'rustup', all the problems were resolved.

**Reply by jpk-frameio** [reply]
Thanks for this!  I had rust and cargo installed via macports, but switched to rustup to fix this.
